### PR TITLE
[Clang][OpenMP] This is addition fix for #92210.

### PR DESCRIPTION
**Pull request by jyu2-git**
Fix another runtime problem when explicit map both pointer and pointee in target data region.

In #92210, problem is only addressed in target region, but missing for target data region.

The change just passing AreBothBasePtrAndPteeMapped in generateInfoForComponentList when processing target data.

### CODE REVIEW

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 1]
```diff
@@ -8025,6 +8025,19 @@ class MappableExprsHandler {
       MapCombinedInfoTy StructBaseCurInfo;
       const Decl *D = Data.first;
       const ValueDecl *VD = cast_or_null<ValueDecl>(D);
+      bool HasMapBasePtr = false;
+      bool HasMapArraySec = false;
+      for (const auto &M : Data.second) {
+        for (const MapInfo &L : M) {
+          const Expr *E = L.VarRef;
+          if (VD && E && VD->getType()->isAnyPointerType() &&
+              isa<DeclRefExpr>(E))
```
Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:8034]:
```suggestion
          if (VD && VD->getType()->isAnyPointerType() &&
              isa_and_present<DeclRefExpr>(E))
```

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 2]
```diff
@@ -8025,6 +8025,19 @@ class MappableExprsHandler {
...
+            HasMapBasePtr = true;
+          if (VD && E && VD->getType()->isAnyPointerType() &&
+              (isa<ArraySectionExpr>(E) || isa<ArraySubscriptExpr>(E)))
```
Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:8037]:
```suggestion
          if (VD && VD->getType()->isAnyPointerType() &&
              isa_and_present<ArraySectionExpr, ArraySubscriptExpr>(E))
```

**Comment by jyu2-git** [comment]
Thanks, Alexey, for the review!

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 3]
```diff
@@ -8025,6 +8025,19 @@ class MappableExprsHandler {
       MapCombinedInfoTy StructBaseCurInfo;
       const Decl *D = Data.first;
       const ValueDecl *VD = cast_or_null<ValueDecl>(D);
+      bool HasMapBasePtr = false;
+      bool HasMapArraySec = false;
+      for (const auto &M : Data.second) {
```
Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:8030]:
```suggestion
      if (VD && VD->getType()->isAnyPointerType()) {
        for (const auto &M : Data.second) {
```

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 4]
```diff
@@ -8025,6 +8025,19 @@ class MappableExprsHandler {
...
+        for (const MapInfo &L : M) {
+          const Expr *E = L.VarRef;
+          if (VD && VD->getType()->isAnyPointerType() &&
+              isa_and_present<DeclRefExpr>(E))
+            HasMapBasePtr = true;
+          if (VD && VD->getType()->isAnyPointerType() &&
+              isa_and_present<ArraySectionExpr, ArraySubscriptExpr>(E))
+            HasMapArraySec = true;
+        }
```
Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:8039]:
```suggestion
  HasMapBasePtr = any_of(M, [](const MapInfo &L) { return isa_and_present<DeclRefExpr>(L.VarRef); });
    HasMapArraySec = any_of(M, [](const MapInfo &L) { return isa_and_present<ArraySectionExpr, ArraySubscriptExpr>(L.VarRef); });
  if (HasMapBasePtr && HasMapArraySec)
    break;
```

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 5]
```diff
@@ -8025,6 +8025,15 @@ class MappableExprsHandler {
...
       const ValueDecl *VD = cast_or_null<ValueDecl>(D);
+      bool HasMapBasePtr = false;
+      bool HasMapArraySec = false;
+      if (VD && VD->getType()->isAnyPointerType()) {
+        for (const auto &M : Data.second) {
+  HasMapBasePtr = any_of(M, [](const MapInfo &L) { return isa_and_present<DeclRefExpr>(L.VarRef); });
+    HasMapArraySec = any_of(M, [](const MapInfo &L) { return isa_and_present<ArraySectionExpr, ArraySubscriptExpr>(L.VarRef); });
+  if (HasMapBasePtr && HasMapArraySec)
+    break;
+      }
```
Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:8036]:
```suggestion
      }
      }
```

Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:8036]: Formatting

Comment (jyu2-git) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp]: Yes, I fixed formatting.  Thanks.

**Review by alexey-bataev — APPROVED** [review]
LG

**Comment by jyu2-git** [comment]
Thanks Alexey!!

**Comment by alexey-bataev** [comment]
Missed one thing, the assignments for boolean flags must be |=, otherwise we may miss some combinations.

**Comment by jyu2-git** [comment]
Hi Alexey,
Thank you take look!

> Missed one thing, the assignments for boolean flags must be |=, otherwise we may miss some combinations.

Are you talking about assignment of following?
HasMapBasePtr = any_of(M, [](c..

Do I need to change to:
HasMapBasePtr |= any_of(M, [](c

Thanks.  
Jennifer

**Comment by alexey-bataev** [comment]
> Hi Alexey,
> 
> Thank you take look!
> 
> 
> 
> > Missed one thing, the assignments for boolean flags must be |=, otherwise we may miss some combinations.
> 
> 
> 
> Are you talking about assignment of following?
> 
> HasMapBasePtr = any_of(M, [](c..
> 
> 
> 
> Do I need to change to:
> 
> HasMapBasePtr |= any_of(M, [](c
> 
> 
> 
> Thanks.  
> 
> Jennifer

Yes, I thought it matches the original code behavior.

**Pull request merged**